src/app/src/main/kotlin/org/vaccineimpact/orderlyweb/controllers/web/WorkflowRunController.kt

```diff
@@ -1,37 +1,96 @@
 import org.vaccineimpact.orderlyweb.db.repositories.OrderlyWebWorkflowRunRepository
 import org.vaccineimpact.orderlyweb.db.repositories.WorkflowRunRepository
+import org.vaccineimpact.orderlyweb.models.WorkflowRun
+import org.vaccineimpact.orderlyweb.models.WorkflowRunRequest
 import org.vaccineimpact.orderlyweb.models.WorkflowRunSummary
 import org.vaccineimpact.orderlyweb.viewmodels.WorkflowRunViewModel
-import org.vaccineimpact.orderlyweb.models.WorkflowRun
+import java.time.Instant
 
 class WorkflowRunController(
     context: ActionContext,
-    private val workflowRunRepository: WorkflowRunRepository
+    private val workflowRunRepository: WorkflowRunRepository,
+    private val orderlyServerAPI: OrderlyServerAPI
 ) : Controller(context)
 {
-    constructor(context: ActionContext) : this(context, OrderlyWebWorkflowRunRepository())
-
-    fun getWorkflowRunDetails(): WorkflowRun
-    {
-        val key = context.params(":key")
-        return workflowRunRepository.getWorkflowRunDetails(key)
-    }
+    constructor(context: ActionContext) : this(
+        context,
+        OrderlyWebWorkflowRunRepository(),
+        OrderlyServer(AppConfig()).throwOnError()
+    )
 
     @Template("run-workflow-page.ftl")
     fun getRunWorkflow(): WorkflowRunViewModel
     {
         return WorkflowRunViewModel(context)
     }
 
+    fun getWorkflowRunDetails(): WorkflowRun
+    {
+        val key = context.params(":key")
+        return workflowRunRepository.getWorkflowRunDetails(key)
+    }
+
     fun getWorkflowRunSummaries(): List<WorkflowRunSummary>
     {
         return workflowRunRepository.getWorkflowRunSummaries(
             context.queryParams("email"),
             context.queryParams("namePrefix")
         )
     }
+
+    internal data class WorkflowRunResponse(
+        @SerializedName(value = "workflow_key")
+        val key: String,
+        val reports: List<String>
+    )
+
+    fun createWorkflowRun(): String
+    {
+        val workflowRunRequest =
+            Serializer.instance.gson.fromJson(context.getRequestBody(), WorkflowRunRequest::class.java)
+
+        val body = Serializer.instance.gson.toJson(
+            listOfNotNull(
+                ("changelog" to workflowRunRequest.changelog).takeIf { it.second != null },
+                ("ref" to workflowRunRequest.gitCommit).takeIf { it.second != null },
+                "reports" to workflowRunRequest.reports.map {
+                    mapOf(
+                        "name" to it.name,
+                        "params" to it.params,
+                        "instance" to workflowRunRequest.instances
+                    )
+                }
+            ).toMap()
+        )
+        val response =
+            orderlyServerAPI.post(
+                "/v1/workflow/run/",
+                body,
+                emptyMap()
+            )
+        val workflowRun = response.data(WorkflowRunResponse::class.java)
+        workflowRunRepository.addWorkflowRun(
```

Comment:
So we don't save the list of running report keys (this is what we get in `reports` in the WorkflowRunReponse?) - these will be re-requested from orderly server by workflow key if we need them later? (Front end will get them back now, but will need to retrieve them on page reload etc)

Reply:
Yes, good question. My thinking is that the list of keys isn't much use in itself - if you're interested in statuses then you'll call that endpoint - which gives exactly the same information as the invocation endpoint (i.e. keys for reports) but with statuses for both the workflow and its reports. The report keys will never outlive the workflow key, and it's much more efficient to get the report statuses as a whole by calling the workflow status endpoint than the individual report endpoints. In short: we could store the list but I can't see us ever using it by itself i.e. iterating through the list and getting the statuses.